crates/burn-tensor/src/tensor/data.rs

```diff
@@ -1,145 +1,522 @@
+        if E::dtype() == self.dtype {
+            Box::new(bytemuck::checked::cast_slice(&self.value).iter().copied())
+        } else {
+            match self.dtype {
+                DType::I8 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &i8| e.elem::<E>()),
+                ),
+                DType::I16 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &i16| e.elem::<E>()),
+                ),
+                DType::I32 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &i32| e.elem::<E>()),
+                ),
+                DType::I64 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &i64| e.elem::<E>()),
+                ),
+                DType::U8 => Box::new(self.value.iter().map(|e| e.elem::<E>())),
+                DType::U32 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &u32| e.elem::<E>()),
+                ),
+                DType::U64 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &u64| e.elem::<E>()),
+                ),
+                DType::BF16 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &bf16| e.elem::<E>()),
+                ),
+                DType::F16 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &f16| e.elem::<E>()),
+                ),
+                DType::F32 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &f32| e.elem::<E>()),
+                ),
+                DType::F64 => Box::new(
+                    bytemuck::checked::cast_slice(&self.value)
+                        .iter()
+                        .map(|e: &f64| e.elem::<E>()),
+                ),
+                // bool is a byte value equal to either 0 or 1
+                DType::Bool => Box::new(self.value.iter().map(|e| e.elem::<E>())),
+            }
+        }
+    }
 
-/// Distribution sampler kind for random value of a tensor.
-pub enum DistributionSamplerKind<E>
-where
-    Standard: rand::distributions::Distribution<E>,
-    E: rand::distributions::uniform::SampleUniform,
-{
-    /// Standard distribution.
-    Standard(rand::distributions::Standard),
+    /// Returns the shape of the tensor data.
+    pub fn shape(&self) -> Vec<usize> {
+        self.shape.to_owned()
+    }
```

Comment:
Could we return a slice instead?

Reply:
That makes me think.. do we want to keep the `shape` vec field public of make it private and have users go through this getter?

I originally added this method for other parts of the codebase that referred to the shape and took a mutable ref to the data but kinda changed things before the review so the method is not necessarily required here. We could also remove it.

Reply:
We could put it public, it's more flexible that way.